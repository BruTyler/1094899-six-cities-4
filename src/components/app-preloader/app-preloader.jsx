import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {getLoadingStatus, getOfflineStatus} from '../../reducer/application/selectors';
import {ActionCreator} from '../../reducer/application/application';
import {getCity, getCities} from '../../reducer/data/selectors.js';
import {Operation as DataOperation} from '../../reducer/data/data.js';
import {Operation as UserOperation} from '../../reducer/user/user.js';
import OfflineScreen from '../offline-screen/offline-screen.jsx';
import App from '../app/app.jsx';

class AppPreloader extends PureComponent {
  constructor(props) {
    super(props);
    this._init();
  }

  _init() {
    const {handleFetchingHotels, handleFetchingAuth, handleLoadingStatus, handleOfflineStatus} = this.props;
    handleLoadingStatus(true);
    const loadTasks = [
      handleFetchingAuth(),
      handleFetchingHotels(),
    ];

    Promise.all(loadTasks)
      .catch(handleOfflineStatus);
  }

  componentDidUpdate() {
    const {isLoading, cityList, handleFirstCity, handleLoadingStatus} = this.props;
    if (isLoading && cityList.length > 0) {
      handleFirstCity(cityList);
      handleLoadingStatus(false);
    }
  }

  render() {
    const {isLoading, isOffline, activeCity} = this.props;

    if (isOffline) {
      return <OfflineScreen />;
    }

    if (isLoading || activeCity === undefined) {
      return null;
    }

    return <App />;
  }
}

AppPreloader.propTypes = {
  handleFetchingAuth: PropTypes.func.isRequired,
  handleFetchingHotels: PropTypes.func.isRequired,
  handleLoadingStatus: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isOffline: PropTypes.bool.isRequired,
  cityList: PropTypes.array.isRequired,
  activeCity: PropTypes.shape(),
  handleFirstCity: PropTypes.func.isRequired,
  handleOfflineStatus: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    cityList: getCities(state),
    activeCity: getCity(state),
    isLoading: getLoadingStatus(state),
    isOffline: getOfflineStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleFetchingHotels() {
    return dispatch(DataOperation.loadCitiesWithApartments());
  },
  handleFetchingAuth() {
    return dispatch(UserOperation.checkAuthorization());
  },
  handleLoadingStatus(newLoadingStatus) {
    dispatch(ActionCreator.changeLoadingStatus(newLoadingStatus));
  },
  handleFirstCity(cityList) {
    dispatch(ActionCreator.changeCity(cityList[0].id));
  },
  handleOfflineStatus() {
    dispatch(ActionCreator.setOffline());
  },
});

export {AppPreloader};
export default connect(mapStateToProps, mapDispatchToProps)(AppPreloader);
