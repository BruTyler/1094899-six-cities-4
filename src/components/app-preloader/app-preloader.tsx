import * as React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {getLoadingStatus, getOfflineStatus} from '../../reducer/application/selectors';
import {ActionCreator} from '../../reducer/application/application';
import {getCity, getCities} from '../../reducer/data/selectors';
import {Operation as DataOperation} from '../../reducer/data/data';
import {Operation as UserOperation} from '../../reducer/user/user';
import OfflineScreen from '../offline-screen/offline-screen';
import App from '../app/app';
import {ErrorStatus} from '../../api';

class AppPreloader extends React.PureComponent {
  constructor(props) {
    super(props);
    this._init();
  }

  _init() {
    const {handleFetchingHotels, handleFetchingAuth, handleLoadingStatus, handleOfflineStatus} = this.props;

    const onErrorCatch = (error) => {
      if (error.response.status !== ErrorStatus.UNAUTHORIZED) {
        handleOfflineStatus();
      }
    };

    handleLoadingStatus(true);
    handleFetchingAuth(onErrorCatch);
    handleFetchingHotels(onErrorCatch);
  }

  componentDidUpdate() {
    const {isLoading, cityList, handleFirstCity, handleLoadingStatus} = this.props;
    if (isLoading && cityList.length > 0) {
      handleFirstCity(cityList);
      handleLoadingStatus(false);
    }
  }

  render() {
    const {isLoading, isOffline, activeCity} = this.props;

    if (isOffline) {
      return <OfflineScreen />;
    }

    if (isLoading || activeCity === undefined) {
      return null;
    }

    return <App />;
  }
}

AppPreloader.propTypes = {
  handleFetchingAuth: PropTypes.func.isRequired,
  handleFetchingHotels: PropTypes.func.isRequired,
  handleLoadingStatus: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isOffline: PropTypes.bool.isRequired,
  cityList: PropTypes.array.isRequired,
  activeCity: PropTypes.shape(),
  handleFirstCity: PropTypes.func.isRequired,
  handleOfflineStatus: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  return {
    cityList: getCities(state),
    activeCity: getCity(state),
    isLoading: getLoadingStatus(state),
    isOffline: getOfflineStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  handleFetchingHotels(onErrorCatch) {
    dispatch(DataOperation.loadCitiesWithApartments())
      .catch(onErrorCatch);
  },
  handleFetchingAuth(onErrorCatch) {
    dispatch(UserOperation.checkAuthorization())
      .catch(onErrorCatch);
  },
  handleLoadingStatus(newLoadingStatus) {
    dispatch(ActionCreator.changeLoadingStatus(newLoadingStatus));
  },
  handleFirstCity(cityList) {
    dispatch(ActionCreator.changeCity(cityList[0].id));
  },
  handleOfflineStatus() {
    dispatch(ActionCreator.setOffline());
  },
});

export {AppPreloader};
export default connect(mapStateToProps, mapDispatchToProps)(AppPreloader);
