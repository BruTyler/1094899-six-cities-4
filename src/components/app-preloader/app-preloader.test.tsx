import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {AppPreloader} from './app-preloader';

const EMPTY_HANDLER = () => null;

describe(`<AppPreloader /> render suit`, () => {
  it(`<AppPreloader /> render offline screen`, () => {
    const generatedTree = renderer.create(
        <AppPreloader
          handleFetchingAuth={EMPTY_HANDLER}
          handleFetchingHotels={EMPTY_HANDLER}
          handleLoadingStatus={EMPTY_HANDLER}
          isLoading={true}
          isOffline={true}
          cityList={[]}
          handleFirstCity={EMPTY_HANDLER}
          handleOfflineStatus={EMPTY_HANDLER}
        />
    ).toJSON();

    expect(generatedTree).toMatchSnapshot();
  });

  it(`<AppPreloader /> render preload (null) screen`, () => {
    const generatedTree = renderer.create(
        <AppPreloader
          handleFetchingAuth={EMPTY_HANDLER}
          handleFetchingHotels={EMPTY_HANDLER}
          handleLoadingStatus={EMPTY_HANDLER}
          isLoading={true}
          isOffline={false}
          cityList={[]}
          activeCity={void 0}
          handleFirstCity={EMPTY_HANDLER}
          handleOfflineStatus={EMPTY_HANDLER}
        />
    ).toJSON();

    expect(generatedTree).toMatchSnapshot();
  });
});
