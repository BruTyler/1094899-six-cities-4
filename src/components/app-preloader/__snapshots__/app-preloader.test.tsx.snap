// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppPreloader /> render suit <AppPreloader /> render offline screen 1`] = `
<div
  className="page page--gray page--main"
>
  <main
    className="page__main page__main--index page__main--index-empty"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="cities"
    >
      <div
        className="cities__places-container cities__places-container--empty container"
      >
        <section
          className="cities__no-places"
        >
          <div
            className="cities__status-wrapper tabs__content"
          >
            <b
              className="cities__status"
            >
              No places to stay available
            </b>
            <p
              className="cities__status-description"
            >
              Server does not not response. Try again later
            </p>
          </div>
        </section>
        <div
          className="cities__right-section"
        />
      </div>
    </div>
  </main>
</div>
`;

exports[`<AppPreloader /> render suit <AppPreloader /> render preload (null) screen 1`] = `null`;
